### PR TITLE
Unified request modification for chained proxies

### PR DESCRIPTION
**Pull request by oxtoacart**

### CODE REVIEW

**Comment by myleshorton** [comment]
Very interesting @oxtoacart -- did you see that data race in CI?

```
5f900d8f-3756-473a-a234-e5b1846c6e6a
==================
WARNING: DATA RACE
Write by goroutine 97:
  github.com/getlantern/balancer.(*dialerHeap).Swap()
      github.com/getlantern/balancer/_obj/strategy.go:105 +0x140
  container/heap.Pop()
      /home/travis/.gimme/versions/go1.6.linux.amd64/src/container/heap/heap.go:63 +0x73
  github.com/getlantern/balancer.(*Balancer).Dial()
      github.com/getlantern/balancer/_obj/balancer.go:99 +0x4f1
  github.com/getlantern/balancer.(*Balancer).Dial-fm()
      github.com/getlantern/flashlight/client/_obj/reverseproxy.go:25 +0x7b
  github.com/getlantern/flashlight/client.(*Client).proxiedDialer.func1()
      github.com/getlantern/flashlight/client/_obj/client.go:251 +0x44f
  net/http.(*Transport).dial()
      /home/travis/.gimme/versions/go1.6.linux.amd64/src/net/http/transport.go:655 +0xd2
  net/http.(*Transport).dialConn()
      /home/travis/.gimme/versions/go1.6.linux.amd64/src/net/http/transport.go:758 +0x276c
  net/http.(*Transport).getConn.func4()
      /home/travis/.gimme/versions/go1.6.linux.amd64/src/net/http/transport.go:702 +0x73
Previous read by goroutine 142:
  github.com/getlantern/balancer.(*dialerHeap).onRequest()
      github.com/getlantern/balancer/_obj/strategy.go:129 +0xb5
  github.com/getlantern/balancer.(*Balancer).OnRequest()
      github.com/getlantern/balancer/_obj/balancer.go:60 +0x60
  github.com/getlantern/balancer.(*Balancer).OnRequest-fm()
      github.com/getlantern/flashlight/client/_obj/reverseproxy.go:27 +0x37
  github.com/getlantern/flashlight/client.(*Client).newReverseProxy.func1()
      github.com/getlantern/flashlight/client/_obj/reverseproxy.go:34 +0xc0
  net/http/httputil.(*ReverseProxy).ServeHTTP()
      /home/travis/.gimme/versions/go1.6.linux.amd64/src/net/http/httputil/reverseproxy.go:177 +0x62e
  github.com/getlantern/flashlight/client.(*Client).ServeHTTP()
      github.com/getlantern/flashlight/client/_obj/handler.go:40 +0x518
  net/http.serverHandler.ServeHTTP()
      /home/travis/.gimme/versions/go1.6.linux.amd64/src/net/http/server.go:2081 +0x206
  net/http.(*conn).serve()
      /home/travis/.gimme/versions/go1.6.linux.amd64/src/net/http/server.go:1472 +0x1565
Goroutine 97 (running) created at:
  net/http.(*Transport).getConn()
      /home/travis/.gimme/versions/go1.6.linux.amd64/src/net/http/transport.go:704 +0x363
  net/http.(*Transport).RoundTrip()
      /home/travis/.gimme/versions/go1.6.linux.amd64/src/net/http/transport.go:307 +0xaa5
  github.com/getlantern/flashlight/client.(*noForwardedForRoundTripper).RoundTrip()
      github.com/getlantern/flashlight/client/_obj/reverseproxy.go:134 +0xc6
  github.com/getlantern/flashlight/client.(*errorRewritingRoundTripper).RoundTrip()
      github.com/getlantern/flashlight/client/_obj/reverseproxy.go:83 +0xb3
  net/http/httputil.(*ReverseProxy).ServeHTTP()
      /home/travis/.gimme/versions/go1.6.linux.amd64/src/net/http/httputil/reverseproxy.go:210 +0xb3d
  github.com/getlantern/flashlight/client.(*Client).ServeHTTP()
      github.com/getlantern/flashlight/client/_obj/handler.go:40 +0x518
  net/http.serverHandler.ServeHTTP()
      /home/travis/.gimme/versions/go1.6.linux.amd64/src/net/http/server.go:2081 +0x206
  net/http.(*conn).serve()
      /home/travis/.gimme/versions/go1.6.linux.amd64/src/net/http/server.go:1472 +0x1565
Goroutine 142 (running) created at:
  net/http.(*Server).Serve()
      /home/travis/.gimme/versions/go1.6.linux.amd64/src/net/http/server.go:2137 +0x4d1
  github.com/getlantern/flashlight/client.(*Client).ListenAndServeHTTP()
      github.com/getlantern/flashlight/client/_obj/client.go:126 +0x77b
  github.com/getlantern/flashlight.Run()
      github.com/getlantern/flashlight/_obj/flashlight.go:147 +0x703
  github.com/getlantern/flashlight/app.(*App).Run.func1()
      github.com/getlantern/flashlight/app/_test/_obj_test/app.go:119 +0x84d
```

**Comment by oxtoacart** [comment]
@myleshorton Yeah, I see the race.  It's odd because it doesn't look new per se, unless I'm missing something.

**Comment by oxtoacart** [comment]
@myleshorton I think I fixed the race, let's see what Travis says.

File: src/github.com/getlantern/balancer/strategy.go [hunk 1]
```diff
@@ -97,9 +100,11 @@ func (s *dialerHeap) Pop() interface{} {
 	return x
 }
 
-func (s *dialerHeap) AuthTokens() (tokens []string) {
+func (s *dialerHeap) onRequest(req *http.Request) (tokens []string) {
 	for _, d := range s.dialers {
-		tokens = append(tokens, d.AuthToken)
+		if d.OnRequest != nil {
+			d.OnRequest(req)
+		}
```
Comment (myleshorton) [inline src/github.com/getlantern/balancer/strategy.go]: Is this still supposed to return the tokens here @oxtoacart?

Comment (myleshorton) [inline src/github.com/getlantern/balancer/strategy.go]: Looks like we can just scrap the return value...surprised go let's that compile actually...

**Comment by oxtoacart** [comment]
@myleshorton tests are good

File: src/github.com/getlantern/balancer/strategy.go (hunk 1 above)
Comment (oxtoacart) [inline src/github.com/getlantern/balancer/strategy.go]: Oh, good catch.  Removed.

Comment (myleshorton) [inline src/github.com/getlantern/balancer/strategy.go]: It also seems like theoretically we should have locking on s.dialers no? I guess Push and Pop aren't called much in practice, but if they were it seems like we'd have a bunch of race issues @oxtoacart (not specific to this PR other than the introduction of iterating in a pretty heavily used portion of the code)

Comment (oxtoacart) [inline src/github.com/getlantern/balancer/strategy.go]: The only place where we push and pop is already protected by a mutex.

**Comment by myleshorton** [comment]
Hey @oxtoacart it seems like we might as well add the lantern version number in a header while we're in here... What do you think?

**Comment by oxtoacart** [comment]
@myleshorton If we add the lantern version number, we should first make a provision on the proxy to strip it, right?

**Comment by myleshorton** [comment]
Yeah I suppose that's right @oxtoacart -- let's do that separately.

**Comment by myleshorton** [comment]
What do you think about the locking on `s.dialers` @oxtoacart?

**Comment by oxtoacart** [comment]
@myleshorton We're effectively already locking on s.Dialers inside of OnRequest (in balancer.go)

**Comment by myleshorton** [comment]
This generally LGTM @oxtoacart. I don't love the two different OnRequest methods at https://github.com/getlantern/lantern/blob/balancecheck/src/github.com/getlantern/flashlight/client/chained.go#L80 and https://github.com/getlantern/lantern/blob/balancecheck/src/github.com/getlantern/flashlight/client/chained.go#L106 but I don't see any obvious way around it that would be useful. Merging!

**Pull request merged**